src/commands/optimize.ts

```diff
@@ -0,0 +1,163 @@
+}
+
+export enum Outputs {
+  TERMINAL='terminal',
+  NEW_FILE='new-file',
+  OVERWRITE='overwrite'
+}
+export default class Optimize extends Command {
+  static description = 'optimize asyncapi specification file';
+  isInteractive = false;
+  optimizations?: Optimizations[];
+  outputMethod?: Outputs;
+
+  static examples: Example[] = [
+    'asyncapi optimize ./asyncapi.yaml',
+    'asyncapi optimize ./asyncapi.yaml --no-tty',
+    'asyncapi optimize ./asyncapi.yaml --optimization=remove-components,reuse-components,move-to-components --no-tty',
+    'asyncapi optimize ./asyncapi.yaml --optimization=remove-components,reuse-components,move-to-components --output=terminal --no-tty',
+  ];
+  
+  static flags = {
+    help: Flags.help({ char: 'h' }),
+    optimization: Flags.string({char: 'p', default: Object.values(Optimizations), options: Object.values(Optimizations), multiple: true, description: 'select the type of optimizations that you want to apply.'}),
+    output: Flags.string({char: 'o', default: Outputs.TERMINAL, options: Object.values(Outputs), description: 'select where you want the output.'}),
+    'no-tty': Flags.boolean({ description: 'do not use an interactive terminal', default: false }),
+  };
+
+  static args = [
+    { name: 'spec-file', description: 'spec path, url, or context-name', required: false },
+  ];
+
+  async run() {
+    const { args, flags } = await this.parse(Optimize); //NOSONAR
+    const filePath = args['spec-file'];
+    const specFile = await load(filePath);
+    this.isInteractive = !flags['no-tty'];
+    this.optimizations = flags.optimization as Optimizations[];
+    this.outputMethod = flags.output as Outputs;
+    const optimizer = new Optimizer(specFile.text());
+    const report = await optimizer.getReport();
+    const isURL = !!specFile.getFileURL();
+
+    if (!(report.moveToComponents?.length || report.removeComponents?.length || report.reuseComponents?.length)) {
+      if (isURL) {
+        this.log(`URL ${specFile.getFileURL()} looks optimized!`);
+      } else {
+        this.log(`File ${specFile.getFilePath()} looks optimized!`);
+      }
+      return;
+    }
+    const isTTY = process.stdout.isTTY;
+    if (this.isInteractive && isTTY) {
+      await this.interactiveRun(report);
+    }
+
+    try {  
+      const optimizedDocument = optimizer.getOptimizedDocument({rules: {
+        moveToComponents: this.optimizations.includes(Optimizations.MOVE_TO_COMPONETS),
+        removeComponents: this.optimizations.includes(Optimizations.REMOVE_COMPONENTS),
+        reuseComponents: this.optimizations.includes(Optimizations.REUSE_COMPONENTS)
+      }, output: Output.YAML});
+      
+      const specPath = specFile.getFilePath();
+      let newPath = '';
+      if (specPath) {
+        const pos = specPath.lastIndexOf('.');
+        newPath = `${specPath.substring(0,pos) }_optimized.${ specPath.substring(pos+1)}`;
+      } else {
+        newPath = 'asyncapi_optimized.yaml';
+      }
```

Comment:
please make sure we generate file called `optimized-asyncapi.yml`

we are trying to make sure autocompletion is supported for IDEs and we added AsyncAPI schema to https://github.com/smoya/schemastore/blob/master/src/api/json/catalog.json#L239-L249. There is some filename matching that enables autocompletion by default.

yes autocompletion do not really work in IDEs yet (soon https://github.com/asyncapi/spec-json-schemas/pull/298) but will 🚀 and we should make sure we promote proper naming of files.